test/testhdf5.c

```diff
@@ -47,46 +47,54 @@ main(int argc, char *argv[])
-    AddTest("id", test_ids, NULL, "User-Created Identifiers", NULL);
-    AddTest("misc", test_misc, cleanup_misc, "Miscellaneous", NULL);
+    AddTest("config", test_configure, NULL, cleanup_configure, NULL, 0, "Configure definitions");
+    AddTest("h5system", test_h5_system, NULL, cleanup_h5_system, NULL, 0, "H5system routines");
+    AddTest("metadata", test_metadata, NULL, cleanup_metadata, NULL, 0, "Encoding/decoding metadata");
+    AddTest("checksum", test_checksum, NULL, cleanup_checksum, NULL, 0, "Checksum algorithm");
+    AddTest("skiplist", test_skiplist, NULL, NULL, NULL, 0, "Skip Lists");
+    AddTest("refstr", test_refstr, NULL, NULL, NULL, 0, "Reference Counted Strings");
+    AddTest("file", test_file, NULL, cleanup_file, NULL, 0, "Low-Level File I/O");
+    AddTest("objects", test_h5o, NULL, cleanup_h5o, NULL, 0, "Generic Object Functions");
+    AddTest("h5s", test_h5s, NULL, cleanup_h5s, NULL, 0, "Dataspaces");
+    AddTest("coords", test_coords, NULL, cleanup_coords, NULL, 0, "Dataspace coordinates");
+    AddTest("sohm", test_sohm, NULL, cleanup_sohm, NULL, 0, "Shared Object Header Messages");
+    AddTest("attr", test_attr, NULL, cleanup_attr, NULL, 0, "Attributes");
+    AddTest("select", test_select, NULL, cleanup_select, NULL, 0, "Selections");
+    AddTest("time", test_time, NULL, cleanup_time, NULL, 0, "Time Datatypes");
+    AddTest("ref_deprec", test_reference_deprec, NULL, cleanup_reference_deprec, NULL, 0,
+            "Deprecated References");
+    AddTest("ref", test_reference, NULL, cleanup_reference, NULL, 0, "References");
+    AddTest("vltypes", test_vltypes, NULL, cleanup_vltypes, NULL, 0, "Variable-Length Datatypes");
+    AddTest("vlstrings", test_vlstrings, NULL, cleanup_vlstrings, NULL, 0, "Variable-Length Strings");
+    AddTest("iterate", test_iterate, NULL, cleanup_iterate, NULL, 0, "Group & Attribute Iteration");
+    AddTest("array", test_array, NULL, cleanup_array, NULL, 0, "Array Datatypes");
+    AddTest("genprop", test_genprop, NULL, cleanup_genprop, NULL, 0, "Generic Properties");
+    AddTest("unicode", test_unicode, NULL, cleanup_unicode, NULL, 0, "UTF-8 Encoding");
+    AddTest("id", test_ids, NULL, NULL, NULL, 0, "User-Created Identifiers");
+    AddTest("misc", test_misc, NULL, cleanup_misc, NULL, 0, "Miscellaneous");
 
     /* Display testing information */
-    TestInfo(argv[0]);
+    TestInfo(stdout);
 
     /* Parse command line arguments */
-    TestParseCmdLine(argc, argv);
+    if (TestParseCmdLine(argc, argv) < 0) {
+        fprintf(stderr, "couldn't parse command-line arguments\n");
+        TestShutdown();
+        exit(EXIT_FAILURE);
+    }
 
     /* Perform requested testing */
     PerformTests();
 
     /* Display test summary, if requested */
     if (GetTestSummary())
-        TestSummary();
+        TestSummary(stdout);
 
     /* Clean up test files, if allowed */
     if (GetTestCleanup() && !getenv(HDF5_NOCLEANUP))
```

Comment:
It seems like test-level cleanup functions are invoked twice here. Once by `PerformTests()` after each test, and then during `TestCleanup`.

Reply:
Yes, this is rather an open question of whether the old behavior is useful in any way. testhdf5 previously waited until all tests have run before cleaning up, which has its own problems even when cleanup callbacks were only handling the removal of temporary testing files. I left that behavior in for now since it's currently harmless, but I'm inclined toward removing the TestCleanup() function entirely and just letting the testing framework do it. `GetTestCleanup()` would probably need to be modified to handle the `HDF5_NOCLEANUP` environment variable.

Reply:
One possibility would be to have separate cleanups for individual tests and for the entire framework.

Reply:
Yes I think at this point it makes sense to add in a callback or two that applies for the whole test program to cover the old way that testhdf5 performed cleanup after all the tests ran. It's also been pointed out that one or more tests have a run-once setup callback.